compiler/rustc_codegen_ssa/src/mir/block.rs

```diff
@@ -311,14 +312,91 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         }
     }
 
+    fn get_expectation(
+        &mut self,
+        bb: mir::BasicBlock,
+        discr: &mir::Operand<'tcx>,
+    ) -> Option<mir::ExpectKind> {
+        // First do a quick test if there are any `expect` intrinsics in the BB.
+
+        let Some(discr) = discr.place() else {
+            return None;
+        };
+
+        let no_expect = self.mir[bb].statements.iter().all(|stmt| {
+            let is_expect =
+                if let mir::StatementKind::Intrinsic(box mir::NonDivergingIntrinsic::Expect(..)) =
+                    stmt.kind
+                {
+                    true
+                } else {
+                    false
+                };
+            !is_expect
+        });
+        if no_expect {
+            return None;
+        }
+
+        // There are `expect` intrinsics in the BB, so we need to do the full analysis.
+
+        // Groups of variables which have the same value
+        let mut groups = Vec::new();
+        // Map from variable to group index
+        let mut map = FxHashMap::<mir::Place<'tcx>, usize>::default();
+
+        for stmt in &self.mir[bb].statements {
+            match stmt.kind {
+                // For `expect` intrinsic, mark the entire group as having the expected value.
+                mir::StatementKind::Intrinsic(box mir::NonDivergingIntrinsic::Expect(
+                    mir::Operand::Copy(ref place),
+                    expect_kind,
+                )) => {
+                    if let Some(index) = map.get(place) {
+                        groups[*index] = Some(expect_kind);
+                    } else {
+                        map.insert(*place, groups.len());
+                        groups.push(Some(expect_kind));
+                    }
+                }
+
+                // For assignments:
+                // - if we understand the RHS, add LHS to the same group
+                // - if we don't understand the RHS, remove LHS from any group,
+                //   so that it doesn't have any expected value
+                mir::StatementKind::Assign(box (ref lhs, mir::Rvalue::Use(ref op))) => {
+                    let Some(rhs) = op.place() else {
+                        map.remove(lhs);
+                        continue;
+                    };
+
+                    if let Some(index) = map.get(&rhs) {
+                        map.insert(*lhs, *index);
+                    } else {
+                        map.insert(rhs, groups.len());
+                        map.insert(*lhs, groups.len());
+                        groups.push(None);
+                    }
+                }
```

Comment:
I think we should define valid uses of the intrinsics. In my opinion it is ok to only support using these as direct operand in `if`, `match`, or one of the logical operators. I also wanted to cover the use case when it is wrapped in an inline function in order to allow export in `std::hints`

So this is a valid use:
```rust
if likely(x) || unlikely(y) { ... }
```

This is not a valid use:
```rust
likely(x);
if x { ... }
```

and this is also not a valid use in my opinion:
```rust
likely(*x);
if *x { ... }
```

We only need to cover the MIR that can be generated in the valid use cases. I may be wrong, but since the intrinsics return a value, I don't expect the optimizer to add any indirect assignments.

If you think it's worth it, I can look at clippy and implement a lint that would identify the invalid uses.

Reply:
MIR opts may twist and bend MIR in many ways. Optimized MIR may not look anything like the original MIR. The only truth is the semantics of the MIR we are actually given.